### PR TITLE
Use `partition_point` for `descendant_at`

### LINKED ISSUE #39: `descendant_at` does not work as expected
The function [`descendant_at`](https://github.com/adclz/auto-lsp/blob/d133723bfbd9150c0ec944b4e9f9cf96844dc167/crates/default/src/db/tracked.rs#L92) does not always find an overlapping node if it exists.

Consider the scenario of two children `2`, `3` of node `1` with the offset between `2` and `3`. In this case `1` would yield `Equals` but `2` would yield `Less`, violating the monotonicity required for the binary search.

Image

---

An alternative approach to find overlapping nodes would be a binary search by starting position followed by a walk up the tree ([example](https://github.com/bachorp/varlink-language-server/blob/8691c13bfcedb4d25ce881fc6d93941650569507/src/capabilities/util.rs#L44)) or a walk down the tree with a binary over the children.

### PR DESCRIPTION
**Pull request by adclz**
Related to #39 

Uses [partition_point](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.partition_point) to fix the behavior of `descendant_at`.
All tests are passing but it's worth adding test cases for the edge cases where the initial issue is triggered.

### CODE REVIEW

**Comment by adclz** [comment]
Actually the more i think about it the more it'd be better to mimic tree-sitter [descendant_at_for_point](https://docs.rs/tree-sitter/latest/tree_sitter/struct.Node.html#method.named_descendant_for_point_range).
`descendant_at` requires a byte offset, which the client doesn't send, it has to be computed via [offset_at](https://github.com/adclz/auto-lsp/blob/3071c1e7e738929e19559edc22f01a635fea3b68/crates/core/src/document/mod.rs#L274). Having a position-based alternative would avoid this extra step.

So it'd be better to have something that accepts an `lsp_types::Position` rather than reinvent the wheel with this offset.

`texter` has a convenient `GridIndex` struct that has `From` implementations for both `tree-sitter::Point` and `lsp_types::Position`,  so i'm thinking about doing something like this:

```rust
pub fn descendant_for_position(
  &self,
  start: GridIndex,
  end: GridIndex,
  text: &texter::Text
) -> Option<&Box<dyn AstNode>> {
    debug_assert!(self.nodes.is_sorted());

    start.normalize(text);
    end.normalize(text);

    let idx = self.nodes.partition_point(|f| {
         let s = f.get_range().start_point;
         (s.row, s.column) <= (start.row, start.col)
     });

     self.nodes[..idx].iter().rev().find(|f| {
         let e = f.get_range().end_point;
         (e.row, e.column) >= (end.row, end.col)
     })

}
``` 

Although i am not sure if a `Text` should be passed here, if not, the user has the responsibility to call `normalize` before using this fn, which is something that can be easily forgotten... 

But the flow would be (inspired by [trunks](https://github.com/airblast-dev/trunkls/tree/main)):
- In: 
  Request -> Position -> Into(GridIndex) -> normalize -> descendant_at_for_position

- Out:
  Position -> Into(GridIndex) -> denormalize

File: crates/default/src/db/tracked.rs [hunk 1]
```diff
@@ -86,25 +86,20 @@ impl ParsedAst {
         self.nodes.first()
     }
 
-    /// Returns the first node that contains the given offset.
-    ///
-    /// This method uses binary search to find the node.
+    /// Returns the deepest node that contains the given offset.
     pub fn descendant_at(&self, offset: usize) -> Option<&Box<dyn AstNode>> {
```
Comment (bachorp) [inline crates/default/src/db/tracked.rs]: The name `descendant` doesn't fit in this case. Opposed to tree-sitter's `descendant_*` this one is not relative to a specific node.

**Comment by bachorp** [comment]
> Although i am not sure if a Text should be passed here, if not, the user has the responsibility to call normalize before using this fn, which is something that can be easily forgotten...

Ideally, the user is not bothered with these encoding shenanigans. Any `lsp::Position` that goes into- or comes out of a public function should be interpreted according to the negotiated encoding. This means that a function such as `pub get_lsp_range: AstNode -> lsp_types::Range` should not exist. It also means that a `texter::GridIndex` should not be user-facing because of the encoding ambiguity (normalized/not normalized).

**Comment by adclz** [comment]
Now that the `Document` module relies on texter's **normalize** and **denormalize** fns:

- `offset_at` is gone, replaced by `normalize_position`, which does the **GridIndex** encoding plumbing and returns an adjusted `Position`.

- `ts_range_to_range` was renamed to `denormalize_range`.

- `descendant_at` was renamed to `descendant_for_position`:  no offset anymore, it takes an extra `&Document` and normalizes a given `Position` internally.

I have not encountered these in my use cases, but I also added `normalize_range`/`denormalize_point` helpers, so an `lsp_types::Range` (a pair of Positions) normalized, or a `tree_sitter::Point `denormalized to the client encoding.
